src/main/java/com/rackspacecloud/metrics/rollup/domain/Metric.java

```diff
@@ -1,118 +1,151 @@
-    public MonitoringSystem monitoringSystem;
-    public Map<String, String> systemMetadata;
-    public String collectionLabel;
-    public String collectionTarget;
-    public Map<String, String> collectionMetadata;
+    String accountType;
+    String account;
+    String device;
+    String deviceLabel;
+    Map<String, String> deviceMetadata;
+    String monitoringSystem;
+    Map<String, String> systemMetadata;
+    String collectionName;
+    String collectionLabel;
+    String collectionTarget;
+    Map<String, String> collectionMetadata;
 
     // InfluxDB fields
     public Map<String, Long> ivalues;
     public Map<String, Double> fvalues;
     public Map<String, String> svalues;
     public Map<String, String> units;
 
+
+
+    public Map<String, RollupBucket<Long>> rolledUpIValues;   // Keeps rolled up ivalue
+    public Map<String, RollupBucket<Double>> rolledUpFValues; // Keeps rolled up fvalue
+
+    @Data
+    public static class RollupBucket<T extends Number> {
+        private T min;
+        private T max;
+        private T mean;
+
+        private T sum;
+        private long count;
+    }
+
     /**
      * Define the rollup-key that is used to aggregate the data.
      * @return rollup-key
      */
     @Override
-    public String getRollupKey(){
-        StringBuilder sb = new StringBuilder();
-        sb.append(String.format("%s", getSystemMetadata()));
-        this.rollupKey = sb.toString();
+    public String getRollupKey() {
+        String[] rollupKeyTags = new String[] {
+                accountType, account, monitoringSystem, collectionName,
+                device, deviceLabel, collectionLabel
+        };
 
-        // TODO: Check what else needs to be part of the rollup key
+        this.rollupKey = String.join(".", rollupKeyTags);
 
-        return rollupKey;
-    }
+        // TODO: Check what else needs to be part of the rollup key
 
-    private String getSystemMetadata(){
-        return String.format("%s.%s.%s.%s.%s",
-                systemMetadata.get("tenantId"),
-                systemMetadata.get("checkType"),
-                systemMetadata.get("accountId"),
-                systemMetadata.get("entityId"),
-                systemMetadata.get("checkId")
-        );
+        return this.rollupKey;
     }
 
     @Override
     public Metric reduce(Metric newValue) {
-        reduceIValues(newValue);
-        reduceFValues(newValue);
+        reduceIValues(newValue.getIvalues());
+        reduceFValues(newValue.getFvalues());
         return this;
     }
 
-    private void reduceIValues(Metric newValue) {
-        if(iValuesForRollup == null) {
-            iValuesForRollup = new HashMap<>();
-            ivalues.forEach((k,v) -> {
-                List<Long> valueList = new ArrayList<>();
-                valueList.add(v);
-                iValuesForRollup.put(k, valueList);
-            });
+    private void reduceIValues(Map<String, Long> newIValues) {
```

Comment:
It looks like this method needs to be `synchronized` since kafka streams threads will call into this.

Reply:
I am using `ConcurrentMap<String, SynchronizedDescriptiveStatistics>` now for thread safety.